### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'bind')

### ISSUE THREAD

**Issue by Aietes** [oftherivier/fictional#39]
I'm having an issue with a setup that worked just fine until now:

```shell
<project>/node_modules/.pnpm/fictional@2.0.2/node_modules/fictional/hash.js:7
var generateKey = siphash.string16_to_key.bind(siphash)
                                          ^

TypeError: Cannot read properties of undefined (reading 'bind')
    at Object.<anonymous> (<project>/node_modules/.pnpm/fictional@2.0.2/node_modules/fictional/hash.js:7:43)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Object.transformer (/Users/xxx/.npm/_npx/fd45a72a545557e9/node_modules/tsx/dist/register-DxMxtvGN.cjs:2:1122)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (<project>/node_modules/.pnpm/fictional@2.0.2/node_modules/fictional/index.js:1:16)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Object.transformer (/Users/xxx/.npm/_npx/fd45a72a545557e9/node_modules/tsx/dist/register-DxMxtvGN.cjs:2:1122)

Node.js v20.15.1
 ELIFECYCLE  Command failed with exit code 1.
```

Can anyone help with this error or point me in the right direction? The error originates in the snaplet/seed library, this is the related issue: https://github.com/snaplet/seed/issues/189

**Comment by josematute** [comment]
Same issue here

**Comment by jmgunter** [comment]
Same issue for me.

**Comment by justinvdm** [comment]
Hi there, I tried reproduce, directly with pnpm+fictional@2.0.2+`npx tsx`, and also via latest `@snaplet/seed` (using pnpm, and a supabase-based db) - no luck just yet.

@Aietes @josematute @jmgunter To help track it down: what version of `@snaplet/copycat` and `@snaplet/seed` are you on?

**Comment by josematute** [comment]
This is what I have in package.json:
...
"@snaplet/copycat": "^5.0.0",
"@snaplet/seed": "^0.97.20",
...

**Comment by justinvdm** [comment]
Thanks @josematute, tried to reproduce on those versions, still no luck.

I'll need your help to get some more details to figure out what's going on.

Could you share the output of this:

```
cd node_modules/fictional && node -e 'console.log(require("siphash"))'
```

**Comment by josematute** [comment]
> cd node_modules/fictional && node -e 'console.log(require("siphash"))'

@justinvdm here it is:

```
{
  hash: [Function: hash],
  hash_hex: [Function: hash_hex],
  hash_uint: [Function: hash_uint],
  string16_to_key: [Function: string16_to_key],
  string_to_u8: [Function: string_to_u8]
}
```

**Comment by justinvdm** [comment]
Thanks @josematute, I'll need to keep digging to try figure this out.

In the meantime, it is a long-shot, but it might be worth upgrading to the latest `@snaplet/seed` ([0.98.0](https://github.com/snaplet/seed/releases/tag/v0.98.0)) to see if this happens to resolve the issue (it comes with breaking changes but the release notes suggest it would need to be done by 31 August 2024 anyways). I can't see how it would be at all related, but fictional, siphash and copycat haven't really changed recently, while `@snaplet/seed` has. Probably a red herring, but maybe there's something there.

**Comment by Aietes** [comment]
I think the problem might be that our projects are defined as a module (ESM) in the package.json by setting "type": "module". Does the module provide both CJS and ESM?

**Comment by ryo-kihara** [comment]
Same issue here

**Comment by jmgunter** [comment]
So oddly enough, deleting the /node_modules/siphash/lib/siphash.ts file fixes the issue for me and lets me run the npx tsx seed.ts file and generate the seed script. Unfortunately removing this file also removes all the types from the seed.ts file so it's not a long term fix but at least lets me run my scripts while we get this sorted out.

**Comment by justinvdm** [comment]
> I think the problem might be that our projects are defined as a module (ESM) in the package.json by setting "type": "module". It looks like the required module siphash is CJS only, and therefore doesn‘t work when used in ESM.
> 
> Switching snaplet/seed versions does not help, I tried various versions earlier today.

I suspected it was something around this too at first, and tried to reproduce along these lines, but didn't have any luck:
* `"type": "module"` in `package.json`
* Writing a `.mjs` script that uses `fictional` directly and calling it with node
* Running a `.ts` script with same content and using `npx tsx`
* Running `npx tsx seed.ts` to test if having `@snaplet/seed` in the mix had anything to do with it

**Comment by justinvdm** [comment]
> So oddly enough, deleting the /node_modules/siphash/lib/siphash.ts file fixes the issue for me and lets me run the npx tsx seed.ts file and generate the seed script. Unfortunately removing this file also removes all the types from the seed.ts file so it's not a long term fix but at least lets me run my scripts while we get this sorted out.

Ah! That gives a thought - it could me related to how the tsconfig is defined, and tsx is looking at that. @jmgunter could you share your tsconfig?

**Comment by jmgunter** [comment]
Ok I just figured it out. The culprit is tsx. running npx tsx@4.16.2 seed.ts works even with the ts file

**Comment by jmgunter** [comment]
> > So oddly enough, deleting the /node_modules/siphash/lib/siphash.ts file fixes the issue for me and lets me run the npx tsx seed.ts file and generate the seed script. Unfortunately removing this file also removes all the types from the seed.ts file so it's not a long term fix but at least lets me run my scripts while we get this sorted out.
> 
> Ah! That gives me an idea - it could me related to how the tsconfig is defined, and tsx is looking at that. @jmgunter could you share your tsconfig?

Sure:

```
{
  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@/*": ["./*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}
```

**Comment by justinvdm** [comment]
Nice find @jmgunter. It looks like this was the relevant change to tsx: https://github.com/privatenumber/tsx/commit/3df00f4e64fd013ef8bc789b0be0496a033f4d3d

It looks like siphash's having a ts export is confusing how tsx resolves after that change. Fortunately we can work around this in fictional, I'll see if I can get that fixed.

In the meantime, you have the workaround you discovered with `npx tsx@4.16.2`.

**Comment by Aietes** [comment]
Thank you for looking into this and getting to a workaround that quickly! Looking forward to a potential longer-term workaround in fictional.

**Comment by camrun91** [comment]
The tsx version stuff fixed it for me as well.

**Comment by justinvdm** [comment]
Hi everyone, I've released fictional 2.0.3 which should fix the issue (https://github.com/oftherivier/fictional/commit/ad5323a3d52197c902e8bd20ec2b53a35809a862).

I was able to reproduce the issue for the case of `@snaplet/seed` using [@jmgunter's tsconfig](https://github.com/oftherivier/fictional/issues/39#issuecomment-2262067283), then confirm that the new fictional version fixed the issue for that case.

For `@snaplet/seed` usage, the new fictional release is in within the version range for latest copycat's fictional dependency, so you won't need to upgrade anything. You will need to get the fictional transitive dep updated in your lockfile though.

Easiest way would probably be to uninstall then reinstall copycat. For example, for yarn:

```
yarn remove @snaplet/copycat@<current_copycat_version_in_your_package_json>
yarn add -D @snaplet/copycat@<current_copycat_version_in_your_package_json>
```

Hope that helps!
